lib/core/test/unit/Cardano/Wallet/Gen.hs

```diff
@@ -290,3 +309,41 @@ genScript elems = scale (`div` 3) $ sized scriptTree
                 , RequireAnyOf scripts'
                 , RequireSomeOf atLeast scripts'
                 ]
+
+genScriptCosigners :: Gen (Script Cosigner)
+genScriptCosigners = do
+    numOfCosigners <- choose (1,10)
+    genScript $ Cosigner <$> [0..numOfCosigners]
+
+genScriptTemplate :: Gen ScriptTemplate
+genScriptTemplate = do
+    script <- genScriptCosigners `suchThat` (not . null . retrieveAllCosigners)
+    let scriptCosigners = retrieveAllCosigners script
+    cosignersSubset <- sublistOf scriptCosigners `suchThat` (not . null)
+    xpubs <- vectorOf (length cosignersSubset) genXPub
+    pure $ ScriptTemplate (Map.fromList $ zip cosignersSubset xpubs) script
+
+genScriptTemplateComplete :: Gen ScriptTemplate
+genScriptTemplateComplete = do
+    script <- genScriptCosigners `suchThat` (not . null . retrieveAllCosigners)
+    let scriptCosigners = retrieveAllCosigners script
+    xpubs <- vectorOf (length scriptCosigners) genXPub
+    pure $ ScriptTemplate (Map.fromList $ zip scriptCosigners xpubs) script
+
+genRootKeySeqWithPass
+    :: Passphrase "encryption"
+    -> Gen (ShelleyKey depth XPrv)
+genRootKeySeqWithPass encryptionPass = do
+    s <- SomeMnemonic <$> genMnemonic @15
+    g <- Just . SomeMnemonic <$> genMnemonic @12
+    return $ Shelley.unsafeGenerateKeyFromSeed (s, g) encryptionPass
+
+genXPub :: Gen XPub
+genXPub = do
+    getRawKey . publicKey <$> (genRootKeySeqWithPass =<< genPassphrase (0, 16))
```

Comment:
Minor note: any sequence of 64 bytes is a valid `XPub`. So for generating an arbitrary key, it would suffice to generate random bytes and use the `xpubFromBytes` smart constructor. No need for true generation from entropy involving mnemonic which is more computationally costly.